src/Utilities/cachingoptimizer.jl

```diff
@@ -393,9 +395,21 @@ end
 
 function MOI.add_constraint(
     m::CachingOptimizer,
-    func::MOI.AbstractFunction,
-    set::MOI.AbstractSet,
-)
+    func::F,
+    set::S,
+) where {F<:MOI.AbstractFunction,S<:MOI.AbstractSet}
+    if m.state != NO_OPTIMIZER && !MOI.supports_constraint(m.optimizer, F, S)
+        reset_optimizer(m)
+        new_opt = if supports_default_copy_to(m.optimizer, false)
```

Comment:
We should create a function doing this to avoid having twice the same code: here and
https://github.com/jump-dev/MathOptInterface.jl/blob/ac22378a3dd310603ced6ccf0141f16d73cae091/src/instantiate.jl#L120-L125

Reply:
Done.